Code/GraphMol/TautomerQuery/TautomerQuery.cpp

```diff
@@ -0,0 +1,259 @@
+    const SubstructMatchParameters &params) const {
+  for (auto idx : d_modifiedAtoms) {
+    const auto queryAtom = tautomer.getAtomWithIdx(idx);
+    const auto targetAtom = mol.getAtomWithIdx(getTargetIdx(idx, match));
+#ifdef VERBOSE
+    std::cout << "Query atom " << queryAtom->getSymbol() << " target atom "
+              << targetAtom->getSymbol() << std::endl;
+#endif
+    if (!atomCompat(queryAtom, targetAtom, params)) {
+#ifdef VERBOSE
+      std::cout << "Atom incompatibility" << std::endl;
+#endif
+      return false;
+    }
+  }
+
+  for (auto idx : d_modifiedBonds) {
+    const auto queryBond = tautomer.getBondWithIdx(idx);
+    const auto beginIdx = queryBond->getBeginAtomIdx();
+    const auto endIdx = queryBond->getEndAtomIdx();
+    const auto targetBeginIdx = getTargetIdx(beginIdx, match);
+    const auto targetEndIdx = getTargetIdx(endIdx, match);
+    const auto targetBond =
+        mol.getBondBetweenAtoms(targetBeginIdx, targetEndIdx);
+#ifdef VERBOSE
+    std::cout << "Query bond " << queryBond->getBondType() << " target bond "
+              << targetBond->getBondType() << std::endl;
+#endif
+    if (!bondCompat(queryBond, targetBond, params)) {
+#ifdef VERBOSE
+      std::cout << "Bond incompatibility" << std::endl;
+#endif
+      return false;
+    }
+  }
+
+#ifdef VERBOSE
+  std::cout << "Tautomer match" << std::endl;
+#endif
+  return true;
+}
+
+std::vector<MatchVectType> TautomerQuery::substructOf(
+    const ROMol &mol, const SubstructMatchParameters &params,
+    std::vector<ROMOL_SPTR> *matchingTautomers) const {
+  if (matchingTautomers) {
+    matchingTautomers->clear();
+  }
+  std::vector<MatchVectType> matches;
+
+#ifdef VERBOSE
+  std::cout << "Tautomer search with query " << MolToSmiles(*d_templateMolecule)
+            << " on " << MolToSmiles(mol) << " max matches "
+            << params.maxMatches << std::endl;
+#endif
+  SubstructMatchParameters templateParams(params);
+  templateParams.maxMatches = 1000;
+  templateParams.uniquify = false;
+  const auto templateMatches =
+      RDKit::SubstructMatch(mol, *d_templateMolecule, templateParams);
+#ifdef VERBOSE
+  std::cout << "Number of template matches " << templateMatches.size()
+            << std::endl;
+#endif
+
+  // TODO create a functor so that I dont have to get all template matches before evaluating tautomer matches.
+  for (auto templateMatch : templateMatches) {
+#ifdef VERBOSE
+    std::cout << "Checking template match" << std::endl;
+#endif
+    for (auto tautomer : d_tautomers) {
+#ifdef VERBOSE
+      std::cout << "Checking Tautomer " << MolToSmiles(*tautomer) << std::endl;
+#endif
+      if (matchTautomer(mol, *tautomer, templateMatch, params)) {
+#ifdef VERBOSE
+        std::cout << "Got Match " << std::endl;
+#endif
+        matches.push_back(templateMatch);
+        if (matchingTautomers) {
+          matchingTautomers->push_back(tautomer);
+        }
+        if (matches.size() == params.maxMatches) goto searchFinished;

```

Comment:
oh no... a goto?
As odd as it is to see one of these, particulary in "modern" C++, I can't think of a more straightforward solution to the problem you're solving. Introducing a new boolean solely for the purpose of avoiding this doesn't make a ton of sense.

Reply:
Yes, I was surprised that "modern" C++ does not have labelled loops and this appears to be the best way to break out of an inner loop.